hazelcast/src/main/java/com/hazelcast/map/impl/recordstore/DefaultRecordStore.java

```diff
@@ -847,10 +839,9 @@ public boolean merge(MapMergeTypes mergingEntry,
             newValue = persistenceEnabledFor(provenance)
                     ? mapDataStore.add(key, newValue, record.getExpirationTime(), now, null) : newValue;
             onStore(record);
-            mutationObserver.onUpdateRecord(key, record, newValue);
+            mutationObserver.onUpdateRecord(key, record, oldValue, newValue, false);
             storage.updateRecordValue(key, record, newValue);
         }
-        saveIndex(record, oldValue);
```

Comment:
@ahmetmircik , @mmedenjak @taburet it seems we've introduced a bug here. In this PR we've refactored the index updates into observes and do the index update BEFORE updating the storage. The update removes the old entry and inserts new one which is exactly the same because it takes it from the `record` that is not updated. After that, we do the storage update and remove the old value. The result is a dangling pointer from the index causing inconsistent results on-heap and JVM crashes off-heap. 

It seems we have to fix it in all affected versions. I'm wondering, do we have a proper split brain test for this?

Reply:
No, I don't remember seeing a test with split brain and indexes.